### ISSUE TITLE
RetryAttribute

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#585]
We accepted this as a feature for NUnit 3.0 but the launchpad bug was never made into a GitHub issue. This has been debated a lot. It has it's down side but we can warn about that in the docs.

Please see https://bugs.launchpad.net/nunit-3.0/+bug/1039415 for details.

**Comment by CharliePoole** [comment]
This will be an easy fix once RepeatAttribute is done because it will look a lot like it in the code. Not quite so easy if it's done first.

**Comment by CharliePoole** [comment]
RepeatAttribute is now done and avaialble as a model for this one.

**Comment by CharliePoole** [comment]
This should be pretty much the same as RepeatAttribute except that it will return on first pass and fail if ii goes through all the repeats without managing to pass. The code will be quite small so trying to create a common base would be a waste of time. Also, it's likely that Repeat will be enhanced in the future in ways that retry will not, so let's keep them separate.

@agray Can I interest you in this one?

**Comment by agray** [comment]
Is this to be addressed in all the nunit.framework-X.0 projects of only one in particular?

Thoughts?

**Comment by CharliePoole** [comment]
Definitely in all. It shouldn't have any requirements that can't be met on all platforms. You'll need to use conservative coding so it can build under VS2008 for the compact framework but I think you'll find that the RepeatAttribute is already coded that way for a guide.

**Comment by agray** [comment]
I have created a new branch for this issue called issue-585 in my fork.  Question:  How does this new branch in my fork find its way back to the main [this] nunit repo?

**Comment by rprouse** [comment]
Once you push your changes up to your branch on GitHub, there should be a
create pull request button at the top of the source on your fork on the web
page.

**Comment by CharliePoole** [comment]
@agray I haven't seen the pull request. Are you still working on this?

**Comment by CharliePoole** [comment]
OK. We're just trying to figure out what to put into the second beta next week. We can push this one to the next release.

**Comment by CharliePoole** [comment]
@agray Any progress on this? If you have not started, I'd like to free it up for someone else to work on.

**Comment by agray** [comment]
Ok, give it to someone else.  Havent been able to crack it yet

**Comment by di97mni** [comment]
:+1: 
RetryAttribute will help us with occasionally failing Selenium tests

**Comment by CharliePoole** [comment]
Here is a mini-spec for this feature as I am implementing it...

RetryAttribute is only allowed on test methods.
Only one attribute is allowed.
Attribute is not inherited.
If the test returns a failure (that is, assertion failure) then it is retried up to the max specified
If it returns anything else (ignored, error, inconclusive, passed) it is not retried.
Note that an unexpected exception is an error, not a failure.

**Comment by di97mni** [comment]
Why not inherited and on class level as well? It would make it easier to retry failing Selenium tests by putting this attribute on a base class

**Comment by CharliePoole** [comment]
@di97mni WRT inheritance, it was left out because other, similar attributes are not inherited. We could consider it if it doesn't appear to cause any conflicts.

WRT class level, I'm not sure how to define it at the class level. What do you do if you run a class and some tests pass while others fail?

**Comment by di97mni** [comment]
Defined on class level would just be an easier way than having to put it on each method. Then if it fails I would asume that the failing methods/tests was run again

**Comment by CharliePoole** [comment]
I see. That would be similar to how we handle Timeout. When placed on a
class, it sets the default for any methods in that class. It's however the
only NUnit attribute that works that way.

Given that many people thought we shouldn't have Retry at all, I don't
think making it easier to apply would get a really good reception. :-)

As you may have noticed, RetryAttribute is now implemented and waiting to
be merged. I think further enhancements may end up waiting for another
go-round.

**Comment by di97mni** [comment]
OK, then we wait for next round.
Thanks for looking into it!

**Comment by di97mni** [comment]
Will there be a new issue to follow for the "next round"
https://github.com/nunit/nunit/issues/585#issuecomment-126988556

**Comment by CharliePoole** [comment]
I don't know. :-)

Seriously, if somebody files a new issue or starts a discussion on nunit-discuss, then we'll talk about it. If I know our community, I suspect some will argue to remove the attribute, while others will want to extend it. 

Based on past discussions, the majority of users think that a retry attribute is just a sign of lazy programming. "Write your tests so they either pass or fail." In practice, we can recognize that sometimes things out of your control require the use of Retry. OTOH, I'm not sure we should make it any easier to use. 

One serious problem is that Retry, like ExpectedException, which we removed, is not very precise in it's application. Any failure of the test method causes a retry. Take a good test with Retry and add Assert.That(2+2, Is.EqualTo(5). The test will be retried up to the maximum number of tries given.

All that said, this is not the place for a discussion that will lead to anything. I'd welcome your starting a discussion elsewhere, preferably where some volume of users will chime in with their opinions.

**Comment by di97mni** [comment]
I agree that generally you shouldn't use Retry but we have tried to fix all occasionally failing Selenium tests but it's _really_ hard. There is a Retry in XUnit as well and the [motivation for that one](https://github.com/giggio/xunit-retry#why-i-wrote-this) is the same as for us.

Anyhow, is there a nightly nuget build I can try or do I need to wait for Beta 4? Any rough plans when that might be available?

Thanks for the feature and all the good work you do!

**Comment by CharliePoole** [comment]
Actually, the Retry is not **in** xUnit, it's an extension. Maybe that's what we should have done. But it's in now and we have other work to do. :-)

We are not doing nightly builds so you would have to either wait or build from source. According to the schedule of milestones on GitHub, we are four or five days late on the beta 4 but only half done. We'll have to make a decision shortly to release or change the schedule. :-)
